src/components/SendTxHandler.js

```diff
@@ -0,0 +1,138 @@
+
+  // Loading message to be shown after return from the server
+  propagatingMessage = t`Propagating transaction to the network.`
+
+  // Success message to be show after tx is sent
+  successMessage = t`Your transaction was sent successfully!`
+
+  /**
+   * miningEstimation {Number} Estimated seconds to complete the job
+   * jobID {String} Mining job ID
+   * loadingMessage {string} Message to be shown to the user while executing the requests
+   */
+  state = {
+    miningEstimation: null,
+    jobID: null,
+    loadingMessage: this.powMessage,
+  }
+
+  componentDidMount = () => {
+    // Start listening for events
+    this.addSendTxEventHandlers();
+    // Start sendTransaction object (submit job)
+    this.props.sendTransaction.start();
+  }
+
+  /**
+   * Create event listeners for all sendTransaction events
+   */
+  addSendTxEventHandlers = () => {
+    this.props.sendTransaction.on('job-submitted', this.updateEstimation);
+    this.props.sendTransaction.on('estimation-updated', this.updateEstimation);
+    this.props.sendTransaction.on('job-done', this.jobDone);
+    this.props.sendTransaction.on('send-success', this.sendSuccess);
+    this.props.sendTransaction.on('send-error', this.sendError);
+  }
+
+  /**
+   * Method executed after the tx send succeeds
+   *
+   * @param {Object} tx Transaction data
+   */
+  sendSuccess = (tx) => {
+    this.setState({ loadingMessage: this.successMessage });
+    if (this.props.onSendSuccess) {
+      this.props.onSendSuccess(tx);
+    }
+  }
+
+  /**
+   * Method executed after the tx send fails
+   *
+   * @param {String} message Error message
+   */
+  sendError = (message) => {
+    this.setState({ errorMessage: `Error: ${message}`});
+    if (this.props.onSendError) {
+      this.props.onSendError(message);
+    }
+  }
+
+  /**
+   * Method executed after the mining job is done
+   *
+   * @param {Object} data Object with jobID
+   */
+  jobDone = (data) => {
+    this.setState({ miningEstimation: null, loadingMessage: this.propagatingMessage });
+  }
+
+  /**
+   * Method executed when the estimation-updated event is received
+   * Update the state with the new estimation
+   *
+   * @param {Object} data Object with jobID and estimation
+   */
+  updateEstimation = (data) => {
+    this.setState({ miningEstimation: data.estimation });
+  }
+
+  render() {
+    const renderMiningEstimation = () => {
+      const estimation = Math.max(Math.ceil(this.state.miningEstimation), MIN_JOB_ESTIMATION);
+      return t`Estimated time: ${estimation}s`;
+    }
+
+    const renderBody = () => {
+      if (this.state.errorMessage) {
+        return (
+          <p className="text-danger mt-3 white-space-pre-wrap">{this.state.errorMessage}</p>
+        )
+      } else {
+        return (
+          <div className="d-flex flex-column">
+            <p>{this.state.loadingMessage}</p>
+            <p>{this.state.miningEstimation && renderMiningEstimation()}</p>
```

Comment:
Should we add an elapsed time?

Reply:
We already have the estimation time going decreasing, another time increasing would be confusing for the user in my opinion.

What do you think? @msbrogli @obiyankenobi @jansegre

Reply:
I agree.

Reply:
I think we could decrease it on every second, and re-adjust when the server updates. This could turnout very simple if we use the estimated time returned to calculated a "willCompleteAt" absolute time, and the frontend only has to always displays the seconds until "willCompleteAt" regardless of what is updating it.

Reply:
As the tx-mining-service always returns a fixed expected time, I agree that we can decrease it on every second. If it reaches zero, we can just write "soon" as suggested by @jansegre .

Reply:
Before we make any changes, we should settle the matter of the mining estimate. If we can't get a reliable number, this is pointless.

Reply:
I agree with @obiyankenobi. We could hide/comment/disable this UI feature and use it when time estimations are at least decent. I don't think we even need to make any calculations, if we launch it without estimates and measure how long it is taking for a few weeks, we can get an idea of the mining time distribution in practice and make a static guess, and improve in the future.

Reply:
I agree with @obiyankenobi and @jansegre. We can disable this UX feature and release a new version. The log will tell us the expected time vs. the total time, and we can use it to improve the estimator later on. When the estimator is good enough, we enable it again.

Reply:
I've removed the time estimation from the UI and when it's good enough we can add again.